CppCoreGuidelines.md

```diff
@@ -3354,10 +3337,31 @@ Alternatively,
 
 Note this is different from the `return move(...)` anti-pattern from [ES.56](#Res-move)
 
+##### Note
+
+For `struct`s where one member represents the success of an operation, adding an explicit conversion to `bool` is helpful:
+
+    struct parse_result {
+        bool success;
+        int value;
+        explicit operator bool() const { return success; }
+    };
+
+    parse_result parse_int(string_view s) { /* ... */ }
+
+    void f()
+    {
+        if (parse_result result = parse_int("123")) { use(result.value); }
```

Comment:
I always assumed the addition of init-statements in if statements in c++17 was to avoid having to write things like this, favoring something more like
`if (parse_result result = parse_int("123"); result.success)`
I think my static analyzer warns about code like `if (parse_result result = parse_int("123"))` because of how easy it is to confuse `=` and `==` in this context. I could be misremembering though and it doesn't complain in the case where it's initialization rather than re-assignment.

I would also say that if one member represents the success then std::optional is likely the better option. Cppreference says "A common use case for optional is the return value of a function that may fail."

Reply:
Personally I don't mind `if (T x = ...)` syntax, and `std::from_chars_result` got an `operator bool` in C++26 to support this kind of test. Static analysis can and should complain about `if (x = ...)` though.

The PR also mentions that in C++17 there is `std::optional` and in C++23 `std::expected` as a better alternative. The note on `bool`-containing `struct`s is more aimed at C++11 and C++14, and somewhat aimed at C++17 and C++20 when failure is not just `bool`, but an error code.